### PR TITLE
444 noxgenerators odata generate keys that are not computed

### LINKED ISSUE #444: [Nox.Generators] [Odata] Generate keys that are not computed
The generated CreateEntityDto Dto, needs to generate a Property for keys that are not computed(including Entity key) , this means all NoxType except DatabaseNumber, DatabaseGuid, and Nuid.

The user when creating this entity needs to set a value on this keys (set this Property has Required in this model)

### PR DESCRIPTION
**Pull request by anna-naumova25**

### CODE REVIEW

File: src/Nox.Generator/Application/Dto/EntityCreateDto.template.cs [hunk 1]
```diff
@@ -13,6 +14,18 @@ namespace {{codeGeneratorState.ApplicationNameSpace }}.Dto;
 /// </summary>
 public partial class {{className}} : {{entity.Name}}UpdateDto
 {
-    // TODO Add Manual Keys to be set by the user
-
+{{- for key in entity.Keys }}
+    {{- if key.Type == "Nuid" || key.Type == "DatabaseNumber" || keyType == "Guid" -}}
```
Comment (rochar) [inline src/Nox.Generator/Application/Dto/EntityCreateDto.template.cs:18]: should be DatabaseGuid , Guid is not autogenerated

File: src/Nox.Generator/Application/EntityMapper.template.cs [hunk 2]
```diff
@@ -28,6 +28,21 @@ public override void MapToEntity({{entity.Name}} entity, Entity entityDefinition
     #pragma warning disable CS0168 // Variable is declared but never used        
         dynamic? noxTypeValue;
     #pragma warning restore CS0168 // Variable is declared but never used
+    {{ for key in entity.Keys -}}
+    {{- if key.Type == "Nuid" || key.Type == "DatabaseNumber" || keyType == "Guid" -}}
```
Comment (rochar) [inline src/Nox.Generator/Application/EntityMapper.template.cs:32]: same here

File: src/Nox.Lib/Factories/NoxTypeFactoryBase.cs [hunk 3]
```diff
@@ -15,8 +15,9 @@ public NoxTypeFactoryBase(NoxSolution solution)
         public virtual T? CreateNoxType(Entity entityDefinition, string propertyName, dynamic? value)
         {
```
Comment (rochar) [inline src/Nox.Lib/Factories/NoxTypeFactoryBase.cs:16]: we should be explicitly and have a method IsKey(keyname) and according to it get the right definition

File: src/Nox.Solution/Models/Domain/Entities/Entity.cs [hunk 4]
```diff
@@ -96,12 +96,27 @@ public NoxSimpleTypeDefinition GetAttributeByName(string entityName)
                 }
             }
         }
-        return _attributesByName[entityName];
+        return _attributesByName.TryGetValue(entityName, out var attribute) ? attribute : null;
```
Comment (rochar) [inline src/Nox.Solution/Models/Domain/Entities/Entity.cs:99]: No need to do this, if we follow the previous comment

File: src/Nox.Solution/Models/Domain/Entities/Entity.cs [hunk 5]
```diff
@@ -96,12 +96,27 @@ public NoxSimpleTypeDefinition GetAttributeByName(string entityName)
...
+    public NoxSimpleTypeDefinition? GetKeyByName(string entityName)
+    {
+        lock (this)
+        {
+            if (_keysByName is null && Keys is not null)
+            {
+                _keysByName = Keys.ToDictionary(key => key.Name, key => key);
+            }
+        }
+        return _keysByName?.TryGetValue(entityName, out var key) == true ? key : null;
```
Comment (rochar) [inline src/Nox.Solution/Models/Domain/Entities/Entity.cs:111]: the same here

**Review by rochar — CHANGES_REQUESTED** [review]
looks good, comments to review
also I want to show you how we do a end to end test on Nox

**Review by alexandervlasenko — APPROVED** [review]
Please, adress the issues raised by Ricardo, otherwise everything is good.

File: src/Nox.Generator/Application/Dto/EntityCreateDto.template.cs (hunk 1 above)
Comment (anna-naumova25) [reply src/Nox.Generator/Application/Dto/EntityCreateDto.template.cs]: Updated

File: src/Nox.Lib/Factories/NoxTypeFactoryBase.cs (hunk 3 above)
Comment (anna-naumova25) [reply src/Nox.Lib/Factories/NoxTypeFactoryBase.cs:16]: Added IsKey and updated the codebase accordingly

**Review by rochar — APPROVED** [review]

**Pull request merged**